pkg/bbgo/order_execution.go

```diff
@@ -332,24 +334,36 @@ func BatchRetryPlaceOrder(ctx context.Context, exchange types.Exchange, errIdx [
 				createdOrders = append(createdOrders, *createdOrder)
 			}
 		}
+
+		errIdx = errIdxNext
 	}
 
+	timeoutCtx, cancelTimeout := context.WithTimeout(ctx, 30*time.Minute)
+	defer cancelTimeout()
+
 	// if we got any error, we should re-iterate the errored orders
-	for len(errIdx) > 0 {
-		time.Sleep(200 * time.Millisecond)
+	coolDownTime := 200 * time.Millisecond
+
+	// set backoff max retries to 101 because https://ja.wikipedia.org/wiki/101%E5%9B%9E%E7%9B%AE%E3%81%AE%E3%83%97%E3%83%AD%E3%83%9D%E3%83%BC%E3%82%BA
+	backoffMaxRetries := uint64(101)
+	for retryRound := 0; len(errIdx) > 0 && retryRound < 10; retryRound++ {
+		// sleep for 200 millisecond between each retry
+		log.Warnf("retry round #%d, cooling down for %s", retryRound+1, coolDownTime)
+		time.Sleep(coolDownTime)
 
-		// allocate a variable for new error index
-		var errIdxNext []int
+		// reset error index since it's a new retry
+		errIdxNext = nil
 
 		// iterate the error index and re-submit the order
-		for _, idx := range errIdx {
+		log.Warnf("starting retry round #%d...", retryRound+1)
+		for _, idx := range errIdxNext {
 			submitOrder := submitOrders[idx]
 
 			op := func() error {
 				// can allocate permanent error backoff.Permanent(err) to stop backoff
-				createdOrder, err2 := exchange.SubmitOrder(ctx, submitOrder)
+				createdOrder, err2 := exchange.SubmitOrder(timeoutCtx, submitOrder)
```

Comment:
The imeout of `timeoutCtx` is 30 minutes. Assuming that the first retrying action takes over 30 minutes, and all the remaining actions will since the timeoutCtx is already timed out.

We could create another timeout ctx in this retrying for loop.

Reply:
yeah I would like to make the whole SubmitOrders including retry is controled under 30m